apex/contrib/csrc/cudnn_gbn/norm_sample.cpp

```diff
@@ -27,448 +27,453 @@
-
-
-        int64_t epsilon_stride[4];
-        generateStrides(epsilon, epsilon_stride, 4, CUDNN_TENSOR_NHWC);
-        auto scalar_tensor_create = [&epsilon_stride, &epsilon](cudnnDataType_t type,
-                                int64_t id) {
-            return cudnn_frontend::TensorBuilder()
-                   .setDim(4, epsilon)
-                   .setStrides(4, epsilon_stride)
-                   .setId(id)
-                   .setAlignment(16)
-                   .setDataType(type)
-                   .setByValue(true)
-                   .build();
-        };
-
-        auto epsilonTensor       = scalar_tensor_create(CUDNN_DATA_DOUBLE, 110);
-        auto expDecayTensor      = scalar_tensor_create(CUDNN_DATA_DOUBLE, 111);
-
-        std::vector<cudnn_frontend::Tensor_v8> peerStatTensors;
-        for (size_t i = 112; i < 112 + peer_devPtrs.size(); ++i) {
-            peerStatTensors.push_back(peer_tensor_create(CUDNN_DATA_FLOAT, i));
-        }
 
-#if (CUDNN_VERSION >= 8500)
-        // Batch normalization
-        cudnnBackendNormMode_t normalizationMode = CUDNN_BATCH_NORM;
-
-        // Forward training
-        cudnnBackendNormFwdPhase_t phase = CUDNN_NORM_FWD_TRAINING;
-
-        //Create a Finalize node
-        auto batch_norm_op = cudnn_frontend::OperationBuilder(CUDNN_BACKEND_OPERATION_NORM_FORWARD_DESCRIPTOR)
-                    .setNormalizationMode(normalizationMode)
-                    .setNormFwdPhase(phase)
-                    .setxDesc(xTensor)
-                    .setScaleAndBias(scaleTensor, biasTensor)
-                    .setPrevRunningMeanAndVar(inMeanTensor, inVarTensor)
-                    .setNextRunningMeanAndVar(outMeanTensor, outVarTensor)
-                    .setSavedMeanAndInvVar(savedMeanTensor, savedInvVarTensor)
-                    .setEpsilonTensor(epsilonTensor)
-                    .setExpDecayFactorTensor(expDecayTensor)
-                    .setPeerStatTensor(peerStatTensors)
-                    .setyDesc(yTensor)
-                    .build();
+// runtime
+cudnn_frontend::ExecutionPlan run_batch_norm_forward(int64_t *tensorDims,
+						     int64_t *perChannelSum,
+						     int64_t *epsilon,
+						     int64_t *peerDims,
+						     cudnnDataType_t data_type) {
+
+  // get the cudnn handle
+  cudnnHandle_t handle = torch::native::getCudnnHandle();
+
+  // Creates the necessary tensor descriptors
+  int64_t tensor_stride[4];
+  int64_t stride[4];
+  int64_t peer_stride[4];
+
+  // NHWC format. GenerateStrides() takes care of this. Howeever, tensor dims should still be NCHW
+  generateStrides(tensorDims, tensor_stride, (int64_t)4, CUDNN_TENSOR_NHWC);
+  generateStrides(peerDims, peer_stride, (int64_t)4, CUDNN_TENSOR_NHWC);
+
+  auto tensor_create = [&tensor_stride, &tensorDims](cudnnDataType_t type,
+  int64_t id) {
+    return cudnn_frontend::TensorBuilder()
+      .setDim(4, tensorDims)
+        .setStrides(4, tensor_stride)
+          .setId(id)
+            .setAlignment(16)
```

Comment:
Manually setting alignment without checking the actual tensor address seems dangerous.

Reply:
the existing code has that too: this is just a refactor of the code which is already present:

https://github.com/NVIDIA/apex/blob/master/apex/contrib/csrc/cudnn_gbn/norm_sample.cpp

This is somewhat urgent since it is fixing a showstopper bug for mlperf hpc 3.0. I am fine with rewriting this but I want to move fast on this. Is there an example of how this should be done?

Reply:
Yes, e.g., https://github.com/pytorch/pytorch/blob/004ff536e87c9586064fb49c4e581f185f3a9d47/aten/src/ATen/native/cudnn/Conv_v8.cpp#L55